sycl/doc/extensions/SubGroupBitmaskGroupFunctions/sub_group_bitmask_group_algorithms.asciidoc

```diff
@@ -0,0 +1,253 @@
+```
+
+## Group_barrier
+New overloads will be required for `sub_group` that takes a bitmask.
+
+```c++
+namespace sycl::ext::oneapi {
+group_barrier(sycl::sub_group, sycl::sub_group_mask mask, memory_scope fence_scope = sub_group::fence_scope);
+} // namespace sycl::ext::oneapi
+```
+
+## New `async_group_copy` function.
+Currently the `nd_item` and `group` classes have the member function
+`async_work_group_copy`. This will be deprecated in favor of a
+`async_group_copy` that will have the group type as a template parameter, much
+like `group_barrier`.
+
+`async_group_copy` methods would be valid for groups `group and `sub_group`.
+For subgroup, a new bitmasked implementation will also be required.
+
+```c++
+namespace sycl::ext::oneapi {
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, size_t srcStride);
+
+template <typename Group, typename dataT>
+device_event async_group_copy(Group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, size_t destStride);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_local_ptr<dataT> dest, sycl::decorated_global_ptr<dataT> src, size_t numElements, size_t srcStride, sycl::sub_group_mask mask);
+
+template <typename dataT>
+device_event async_group_copy(sycl::sub_group group, sycl::decorated_global_ptr<dataT> dest, sycl::decorated_local_ptr<dataT> src, size_t numElements, size_t destStride, sycl::sub_group_mask mask);
+} // namespace sycl::ext::oneapi
+```
+
+## Change to `device_event::wait` function
+`device_event::wait` will be modified to take the group that will be synchonised in the wait.
+This will be valid for the same groups as `async_group_copy` i.e. `group` or `sub_group`.
+An overload with a `sub_group_mask` parameter will be required for `sub_group`.
+
+```c++
+namespace sycl {
+  class device_event {
+  ...
+  public:
+    template <typename Group>
+    void ext_oneapi_wait(Group group) noexcept;
+    void ext_oneapi_wait(sycl::sub_group group, sycl::sub_group_mask mask) noexcept;
```

Comment:
Bikeshedding: There seem to be two ways to think about what's going on here, and I think we should think carefully about what we name the `event` and this `wait` function to make sure it matches how we want programmers to think.

For example:
- If the event represents the whole group operation and it's only meaningful to wait with the same group that used the algorithm, I'd propose `group_event.wait()` (where the group was stored in the event by the algorithm).
- If the event can represent other (non-group) operations and could be waited on by a single work-item or a whole group, I'd propose `event.group_wait(Group g)` (to distinguish from `event.wait()`.

Reply:
I think that this https://github.com/intel/llvm/pull/4903#discussion_r743803534 applies equally to this comment.